dev-tools/mage/kuberemote.go

```diff
@@ -0,0 +1,612 @@
+}
+
+// syncServiceAccount syncs required service account.
+func (r *KubeRemote) syncServiceAccount() error {
+	// delete before create
+	r.deleteServiceAccount()
+	_, err := r.cs.CoreV1().ServiceAccounts(r.namespace).Create(createServiceAccountManifest(r.svcAccName))
+	if err != nil {
+		return errors.Wrap(err, "failed to create service account")
+	}
+	_, err = r.cs.RbacV1().ClusterRoles().Create(createClusterRoleManifest(r.name))
+	if err != nil {
+		return errors.Wrap(err, "failed to create cluster role")
+	}
+	_, err = r.cs.RbacV1().ClusterRoleBindings().Create(createClusterRoleBindingManifest(r.name, r.namespace, r.svcAccName))
+	if err != nil {
+		return errors.Wrap(err, "failed to create cluster role binding")
+	}
+	return nil
+}
+
+// createPod creates the pod.
+func (r *KubeRemote) createPod(env map[string]string, cmd ...string) (*apiv1.Pod, error) {
+	r.deletePod() // ensure it doesn't already exist
+	return r.cs.CoreV1().Pods(r.namespace).Create(createPodManifest(r.name, "golang:1.13.9", env, cmd, r.workDir, r.destDir, r.secretName, r.svcAccName))
+}
+
+// deletePod deletes the pod.
+func (r *KubeRemote) deletePod() {
+	_ = r.cs.CoreV1().Pods(r.namespace).Delete(r.name, &metav1.DeleteOptions{})
+}
+
+// waitForPod waits for the created pod to match the given condition.
+func (r *KubeRemote) waitForPod(wait time.Duration, condition watchtools.ConditionFunc) (*apiv1.Pod, error) {
+	w, err := r.cs.CoreV1().Pods(r.namespace).Watch(metav1.SingleObject(metav1.ObjectMeta{Name: r.name}))
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, _ := watchtools.ContextWithOptionalTimeout(context.Background(), wait)
+	ev, err := watchtools.UntilWithoutRetry(ctx, w, func(ev watch.Event) (bool, error) {
+		return condition(ev)
+	})
+	if ev != nil {
+		return ev.Object.(*apiv1.Pod), err
+	}
+	return nil, err
+}
+
+// portFoward runs the port forwarding so SSH rsync can be ran into the pod.
+func (r *KubeRemote) portForward(ports []string, stopChannel, readyChannel chan struct{}, stdout, stderr io.Writer) (*portforward.PortForwarder, error) {
+	roundTripper, upgrader, err := spdy.RoundTripperFor(r.cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s/portforward", r.namespace, r.name)
+	hostIP := strings.TrimLeft(r.cfg.Host, "https://")
+	serverURL := url.URL{Scheme: "https", Path: path, Host: hostIP}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, &serverURL)
+	return portforward.New(dialer, ports, stopChannel, readyChannel, stdout, stderr)
+}
+
+// rsync performs the rsync of sync directory to destination directory inside of the pod.
+func (r *KubeRemote) rsync(port uint16, stdout, stderr io.Writer) error {
+	privateKeyFile, err := createTempFile(r.privateKey)
+	if err != nil {
+		return err
+	}
+
+	rsh := fmt.Sprintf("/usr/bin/ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR -p %d -i %s", port, privateKeyFile)
```

Comment:
Does the path to `ssh` need to be hardcoded? Any chance this will work for Windows developers?

Reply:
It does not, I removed the hardcode path. If on windows they had `rsync` and `ssh` in PATH then it most likely would work.